src/differential_privacy/protect_grouping_keys.rs

```diff
@@ -67,51 +30,99 @@ impl Reduce {
-        let on: Vec<Expr> = grouping_values
-            .schema()
-            .iter()
-            .map(|f| {
-                Expr::eq(
-                    Expr::qcol(self.name().to_string(), f.name().to_string()),
-                    Expr::qcol(grouping_values.name().to_string(), f.name().to_string()),
+    pub fn protect_grouping_keys(
+        self,
+        protected_entity_id: &str,
+        protected_entity_weight: &str,
+        epsilon: f64,
+        delta: f64,
+    ) -> Result<(PEPRelation, PrivateQuery)> {
+        Ok(
+            if self.grouping_columns()? == vec![protected_entity_id.to_string()] {
+                (
+                    PEPRelation::try_from(Relation::from(self.clone()))?,
+                    PrivateQuery::null(),
                 )
-            })
-            .collect();
-
-        let right = Relation::from(self.with_grouping_columns());
-        let right_names = right
-            .schema()
-            .iter()
-            .map(|f| f.name().to_string())
-            .collect::<Vec<_>>();
+            } else {
+                let (grouping_values, private_query) = self
+                    .grouping_values(protected_entity_id, protected_entity_weight, epsilon, delta)?
+                    .into();
+                let input_relation_with_protected_grouping_keys = self
+                    .input()
+                    .clone()
+                    .join_with_grouping_values(grouping_values)?;
+                let relation: Relation = Reduce::builder()
+                    .with(self.clone())
+                    .input(input_relation_with_protected_grouping_keys)
+                    .build();
+                (PEPRelation::try_from(relation)?, private_query)
+            },
+        )
+    }
+}
 
-        let join_rel: Relation = Relation::join()
-            .right(right)
-            .right_names(right_names.clone())
-            .left(grouping_values)
-            .left_outer()
-            .on_iter(on)
+impl Map {
+    pub fn protect_grouping_keys(
+        self,
+        epsilon: f64,
+        delta: f64,
+    ) -> Result<(PEPRelation, PrivateQuery)> {
+        let (protected_input, private_query) = PEPRelation::try_from(self.inputs()[0].clone())?
+            .protect_grouping_keys(epsilon, delta)?;
+        let relation: Relation = Map::builder()
+            .with(self.clone())
+            .input(Relation::from(protected_input))
             .build();
+        Ok((PEPRelation::try_from(relation)?, private_query))
+    }
+}
 
-        Ok(join_rel.filter_fields(|f| right_names.contains(&f.to_string())))
+impl Join {
+    pub fn protect_grouping_keys(
+        self,
+        epsilon: f64,
+        delta: f64,
+    ) -> Result<(PEPRelation, PrivateQuery)> {
+        let (left_dp_relation, left_private_query) =
+            PEPRelation::try_from(self.inputs()[0].clone())?
+                .protect_grouping_keys(epsilon, delta)?;
```

Comment:
Are you sure you are not processing the same relation twice?
Can you add a test for this case ?